### PR TITLE
Reproducing the compatibility problem between JdbcIO.readRows() and AvroUtils.toAvroSchema() : (#27380)

### PR DESCRIPTION
**Pull request by mkhludnev**
- AvroUtils can't create schema from Beam schema made by JdbcIO.readRows() for NUMERIC and DECIMAL types.
 - reproducer for (#27380)

Results seem really sad. AvroUtils.toAvroSchema() fails on NUMERIC and DECIMAL (here it pass for INT but I'm gust verifying the test approach). Failure looks like:

```
java.lang.RuntimeException: Unhandled logical type beam:logical_type:fixed_decimal:v1
	at org.apache.beam.sdk.extensions.avro.schemas.utils.AvroUtils.getFieldSchema(AvroUtils.java:1095)
	at org.apache.beam.sdk.extensions.avro.schemas.utils.AvroUtils.toAvroField(AvroUtils.java:453)
	at org.apache.beam.sdk.extensions.avro.schemas.utils.AvroUtils.toAvroSchema(AvroUtils.java:498)
```

### CODE REVIEW

**Comment by mkhludnev** [comment]
My IT hasn't picked by CI, I don't know why. But the failure looks like https://paste.apache.org/wms63

**Comment by mkhludnev** [comment]
> Failure looks like:
> 
> ```
> java.lang.RuntimeException: Unhandled logical type beam:logical_type:fixed_decimal:v1
> 	at org.apache.beam.sdk.extensions.avro.schemas.utils.AvroUtils.getFieldSchema(AvroUtils.java:1095)
> 	at org.apache.beam.sdk.extensions.avro.schemas.utils.AvroUtils.toAvroField(AvroUtils.java:453)
> 	at org.apache.beam.sdk.extensions.avro.schemas.utils.AvroUtils.toAvroSchema(AvroUtils.java:498)
> ```

This particular one can be fixed via https://github.com/apache/beam/blame/2d75860eb83feb167e5f909a2fa004d68649100c/sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java#L1195 but it's just a top of the hill

**Comment by Abacn** [comment]
Thanks, if I understood correctly, the problem is on AvroUtil side; JdbcIO can still read from jdbc and send correct elements (Decimal with correct unit) downstream; but if try to write to Avro files, avro schema cannot be created from Beam schema that Beam Row provides. Is that the issue?

**Comment by mkhludnev** [comment]
> Thanks, if I understood correctly, the problem is on AvroUtil side; JdbcIO can still read from jdbc and send correct elements (Decimal with correct unit) downstream; but if try to write to Avro files, avro schema cannot be created from Beam schema that Beam Row provides. Is that the issue?

1. @Abacn, thanks for looking into. 
2. Your interpretation of end-user problem is correct. 
3. This problem/failure also hides another the smaller one - JdbcIO.readRows() doesn't pass the scale when converts DECIMAL(10,2). 
4.  Can you give a clue where start to tackle this problem? I suppose to add a condition for capturing FixedPrecisionNumeric into `AvroUtils.getFieldSchema(AvroUtils.java:1095)`. WDYT?

**Comment by mkhludnev** [comment]
pushed https://github.com/apache/beam/pull/27406/commits/dd74e7a8aa5ac393fbeeda36d782ff9dd04230ab which just let to convert beam schema to Avro's. Now I'm checking writing Avro for these types.

File: sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java [hunk 1]
```diff
@@ -1055,45 +1049,7 @@ private static org.apache.avro.Schema getFieldSchema(
...
-                      .collect(Collectors.toList()));
-        } else if ("DATE".equals(identifier) || SqlTypes.DATE.getIdentifier().equals(identifier)) {
-          baseType = LogicalTypes.date().addToSchema(org.apache.avro.Schema.create(Type.INT));
-        } else if ("TIME".equals(identifier)) {
-          baseType = LogicalTypes.timeMillis().addToSchema(org.apache.avro.Schema.create(Type.INT));
-        } else {
-          throw new RuntimeException(
-              "Unhandled logical type " + fieldType.getLogicalType().getIdentifier());
-        }
+        baseType = getLogicalTypeFieldSchema(fieldType, fieldName, namespace);
```
Comment (mkhludnev) [inline sdks/java/extensions/avro/src/main/java/org/apache/beam/sdk/extensions/avro/schemas/utils/AvroUtils.java:1052]: Extracted logical type handling before adding FixedPrecisionNumeric case.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/FixedPrecisionNumeric.java [hunk 2]
```diff
@@ -113,4 +115,13 @@ public BigDecimal toInputType(BigDecimal base) {
     }
     return base;
   }
+
+  /**
+   * @TODO is byte or fixed?
+   * @TODO should is pass namespace/field name, etc?
+   * */
+  public org.apache.avro.Schema toAvroType() {
```
Comment (Abacn) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/FixedPrecisionNumeric.java:123]: We are in the process of getting rid of avro dependency in java core (CC: @aromanenko-dev) this function probably should goes to AvroUtils

**Comment by Abacn** [comment]
> I suppose to add a condition for capturing FixedPrecisionNumeric into AvroUtils.getFieldSchema(AvroUtils.java:1095).

Thanks, these sounds good. 

>  JdbcIO.readRows() doesn't pass the scale when converts DECIMAL(10,2).

From the change of this PR, is it due to Jdbc's SchemaUtil treat NUMERIC and DECIMAL differently (NUMERIC -> fixed_numeric logical type; DECIMAL -> DECIMAL field type). There is slightly difference between SQL NUMERIC and DECIMAL. Probably check that did this difference get preserved in the old implementation?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/FixedPrecisionNumeric.java (hunk 2 above)
Comment (mkhludnev) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes/FixedPrecisionNumeric.java:123]: Fair. It needs to access `precision, scale)` which are private, yet, with no getters.

**Comment by mkhludnev** [comment]
> . There is slightly difference between SQL NUMERIC and DECIMAL. Probably check that did this difference get preserved in the old implementation?

Hm.. I don't know. Here's what I've got https://www.postgresql.org/docs/current/datatype-numeric.html#DATATYPE-INT

> The types `decimal` and `numeric` are equivalent. Both types are part of the SQL standard.

**Comment by mkhludnev** [comment]
[7a80a24](https://github.com/apache/beam/pull/27406/commits/7a80a2403b9310260f22f33539dfe4f78ebb8f8f) Adding toGeneric conversion into test pipeline. I have no idea how to fix it. 
I really appreciate if you give me a clue!

**Comment by aromanenko-dev** [comment]
@mkhludnev Thanks for you contribution!
I'm sorry but I'm a bit lost with the current state of your JdbcIO/Avro changes. 

Could you elaborate a bit about that and what are the next steps from your point of view and what help do you need?

**Comment by mkhludnev** [comment]
> Could you elaborate a bit about that and what are the next steps from your point of view and what help do you need?

when Postgres `NUMERIC|DECIMAL` (no precision, no scale) writes value with fractional part `1.23` into Avro.
JdbcIOToAvroIT fails with 
```
Caused by: org.apache.avro.AvroTypeException: Cannot encode decimal with scale 2 as scale 0
	at app//org.apache.avro.Conversions$DecimalConversion.toBytes(Conversions.java:92)
	at app//org.apache.beam.sdk.extensions.avro.schemas.utils.AvroUtils.genericFromBeamField(AvroUtils.java:1162)
```
The problem here is that Avro schema should cary two values for this types: bigdecimal bytes (we have it now) and length of fractional part ie scale (since it might be arbitrary for SQL `NUMERIC`). Or we already have an approach to cary SQL `NUMERIC`  in Avro.

**Comment by mkhludnev** [comment]
> Could you elaborate a bit about that and what are the next steps from your point of view and what help do you need?
When JdbcIOToAvroIT `NUMERIC/DECIMAL(10,2)` (with precision and scale) writes value into Avro. It fails with 
 ```
Caused by: java.lang.RuntimeException: Unhandled logical type beam:logical_type:fixed_decimal:v1
	at org.apache.beam.sdk.extensions.avro.schemas.utils.AvroUtils.genericFromBeamField(AvroUtils.java:1228)
```
`BigDecimal` arrives into `AU.genericFromBeamField` with FixedPrecisionNumber logical type. I'm able to capture it via
 `if (FixedPrecisionNumber.IDENTIFIER.equals(identifier)) {` but what's the GenericData in this case. Should we just yield same BigDecimal or I should somehow construct bytes of FixedSchema?

**Comment by aromanenko-dev** [comment]
@mkhludnev Thanks! Sorry, I didn't have a time this week to take a look more deeply on this isse and the next week I'll be mostly off. So, we can come back to this only after.

In the same time, since seems there is no obvious solution, I'd recommend to start a discussion on `dev@beam.apache.org` and explain this issue in details there to have more eyes in this problem.

**Pull request closed without merging**